In [4]:
#Implementation from Scratch Step - by - Step
#To - Do - 1:
import pandas as pd
dataset_path = "/content/drive/MyDrive/Concepts of AI /Datasets/student.csv"
data = pd.read_csv(dataset_path)

#2
print("Top 5 rows of the dataset:")
print(data.head())
print("\nBottom 5 rows of the dataset:")
print(data.tail())

#3
print("\nDataset Information:")
data.info()

#4
print("\nDescriptive statistics of the dataset:")
print(data.describe())

#5
X = data[['Reading', 'Writing']]
Y = data['Math']
print("\nFeatures(X): \n",X.head())
print("\nTarget(Y): \n",Y.head())

Top 5 rows of the dataset:
   Math  Reading  Writing
0    48       68       63
1    62       81       72
2    79       80       78
3    76       83       79
4    59       64       62

Bottom 5 rows of the dataset:
     Math  Reading  Writing
995    72       74       70
996    73       86       90
997    89       87       94
998    83       82       78
999    66       66       72

Dataset Information:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column   Non-Null Count  Dtype
---  ------   --------------  -----
 0   Math     1000 non-null   int64
 1   Reading  1000 non-null   int64
 2   Writing  1000 non-null   int64
dtypes: int64(3)
memory usage: 23.6 KB

Descriptive statistics of the dataset:
              Math      Reading      Writing
count  1000.000000  1000.000000  1000.000000
mean     67.290000    69.872000    68.616000
std      15.085008    14.657027    15.241287
min      13.000000    19.000000    14.000000
25%    

In [7]:
#To - Do - 2:
import numpy as np
import pandas as pd
data = pd.read_csv("/content/drive/MyDrive/Concepts of AI /Datasets/student.csv" )  # Update the path if needed

#2
X = data[["Math"]].to_numpy().T

#3
d = X.shape[0]
W = np.random.rand(d, 1)

#4
Y = np.dot(W.T, X)

#5
print("Feature Matrix (X):")
print(X)

print("\nWeight Matrix (W):")
print(W)

print("\nOutput Matrix (Y):")
print(Y)

Feature Matrix (X):
[[ 48  62  79  76  59  69  70  46  61  86  62  72  56  81  61  49  60  45
   71  75  66  57  67  63  68  68  62  56  82  64  71  69  64  76  61  47
   50  75  69  42  73  78  65  51  61  69  51  64  90  58 100  73  62  45
   47  53  62  49  77  70  60  82 100  72  62  87  56  96  68  66  68  59
   52  60  60  46  62  30  83  69  82  64  57  71  37  69  72  73  70  75
   54  71  60  81  58  54  49  64  84  84  33  51  68  32  43  73  82  47
   74  36  58  68  92  75  43  68  49  69  91  83  65  79  74  81  41  48
   31  53  87  94  77  57  64  55  61  78  94  88  77  90  55  83  68  66
   56  65  84  90  80  47  67  79  68  62  62  44  74  67  46  71  73  49
   42  55  73  75  96  66  56 100  53  68  72  55  59  76  84  68  61  53
   73  61  73  63  51  70  69  61  88  90  60 100  68  85  41  70  91  68
   63  81  89  66  59  80  87  30  63  43  73  49  65  72  83  46  65 100
   40  73  58  87  61  71  72  81  67  69  53  70  56  84  44  63  82  61
   68  75  78  32 

In [9]:
#To - Do - 3:
import pandas as pd
data = pd.read_csv("/content/drive/MyDrive/Concepts of AI /Datasets/student.csv")

#1
train_ratio = 0.8
train_size = int(len(data) * train_ratio)

train_data = data.iloc[:train_size]
test_data = data.iloc[train_size:]

#2
X_train = train_data.drop('Math', axis=1)
y_train = train_data['Math']
X_test = test_data.drop('Math', axis=1)
y_test = test_data['Math']
print(f"Size of Training Data: {len(X_train)}")
print(f"Size of Test Data: {len(X_test)}")

Size of Training Data: 800
Size of Test Data: 200


In [12]:
#To - Do:
import numpy as np
import pandas as pd

def gradient_descent(X, y, W, learning_rate, num_iterations):
    m = len(y)
    for _ in range(num_iterations):
        predictions = X.dot(W)
        errors = predictions - y
        gradient = (1 / m) * X.T.dot(errors)
        W -= learning_rate * gradient
    return W

def rmse(y_true, y_pred):
    return np.sqrt(np.mean((y_true - y_pred) ** 2))

def r2_score_custom(y_true, y_pred):
    total_variance = np.sum((y_true - np.mean(y_true)) ** 2)
    unexplained_variance = np.sum((y_true - y_pred) ** 2)
    return 1 - (unexplained_variance / total_variance)

def train_and_evaluate_model(data, target_column, learning_rate=0.01, num_iterations=1000, train_ratio=0.8):
    X = data.drop(columns=[target_column]).values
    y = data[target_column].values.reshape(-1, 1)
    X = (X - X.mean(axis=0)) / X.std(axis=0)
    X = np.c_[np.ones((X.shape[0], 1)), X]
    train_size = int(len(data) * train_ratio)
    X_train, X_test = X[:train_size], X[train_size:]
    y_train, y_test = y[:train_size], y[train_size:]
    W = np.zeros((X_train.shape[1], 1))
    W = gradient_descent(X_train, y_train, W, learning_rate, num_iterations)
    y_pred = X_test.dot(W)
    rmse_value = rmse(y_test, y_pred)
    r2_value = r2_score_custom(y_test, y_pred)
    print(f"RMSE: {rmse_value:.4f}")
    print(f"R²: {r2_value:.4f}")
    return W

data = pd.read_csv("/content/drive/MyDrive/Concepts of AI /Datasets/student.csv")
W = train_and_evaluate_model(data, target_column='Math', learning_rate=0.01, num_iterations=1000)

RMSE: 8.0630
R²: 0.6688


An R² value of 0.67 indicates that the model has moderate predictive ability. While it is not flawless, it offers a reasonable approximation of the data.

The RMSE of 8.06 reflects a moderate level of error between the predicted and actual values. However, without understanding the range or scale of the target variable, it’s challenging to determine whether this level of error is significant or acceptable.

In [15]:
#2. experiment with different value of learning rate
learning_rates = [0.001, 0.01, 0.1, 0.5, 1]
for lr in learning_rates:
    print(f"Evaluating model with learning rate: {lr}")
    W = train_and_evaluate_model(data, target_column='Math', learning_rate=lr, num_iterations=1000)
    print("-" * 50)

Evaluating model with learning rate: 0.001
RMSE: 26.4877
R²: -2.5742
--------------------------------------------------
Evaluating model with learning rate: 0.01
RMSE: 8.0630
R²: 0.6688
--------------------------------------------------
Evaluating model with learning rate: 0.1
RMSE: 8.0551
R²: 0.6694
--------------------------------------------------
Evaluating model with learning rate: 0.5
RMSE: 8.0551
R²: 0.6695
--------------------------------------------------
Evaluating model with learning rate: 1
RMSE: 694744374.0630
R²: -2458907675230898.0000
--------------------------------------------------


Low Learning Rate (e.g., 0.001): Training will be slower, but the model is less likely to overshoot and can converge more steadily. However, if the model performs poorly, it may be because it hasn't learned enough due to the slow updates.

Moderate Learning Rate (e.g., 0.01): Offers a balance between speed and convergence, making it a good starting point for many models.

High Learning Rate (e.g., 0.1, 0.5, 1): Training may be faster, but there's a risk of overshooting the optimal weights. If the learning rate is too high, the model might not converge or could result in large errors (high RMSE).